stdlib/3/typing.pyi

```diff
@@ -663,3 +663,15 @@ if sys.version_info >= (3, 7):
         def __eq__(self, other: Any) -> bool: ...
         def __hash__(self) -> int: ...
         def __repr__(self) -> str: ...
+
+class _NoneType:
+    __doc__: Optional[str]
+    def __bool__(self) -> bool: ...
```

Comment:
The return type of `__bool__` could be `Literal[False]`.

Reply:
If we do this, we'd need to make it conditional on , since `Literal` isn't defined in `typing.pyi` prior to 3.8, and I don't think we want to import `typing_extensions` from `typing`.

Reply:
This could be solved by @hauntsaninja's suggestion above to move it to `_typeshed`. But I assume you want to special-case it in pyright and would prefer to have it in `typing`?

Reply:
Yeah, I have a slight preference for having it in `typing`, but if the consensus is that it belongs in `_typeshed.pyi`, I could just as easily special-case that.

Reply:
Would it make sense to put this in a `pyright_extensions` module that pyright itself distributes with stubs? That's what we've done in the past with mypy- and pyre-specific primitives.

It would be more work to maintain though; I'm also fine with putting it here or in `_typeshed`.

Does it also need to be in `stdlib/2/typing.pyi`?

Reply:
I think this would be useful for all type checkers though. If it's not, then I'll just hard-code it within pyright.

Reply:
If we feel this is a type that could be generally useful, we should put it into `_typeshed`. This also resolves questions like stdlib/2/typing.pyi and use of Literal. (And if we don't feel that way, we should just push for pyright to hardcode)
Undocumented non-existent types in typing / builtins are hard to discover / when discovered sometimes cause confusion, eg AwaitableGenerator.

Reply:
Especially if this class has the special property of basically not inheriting from object, putting it in a pyright-specific location such as `pyright_extensions` seem like the best option for me. Otherwise we'd be introducing this new concept to typeshed that other type checkers (at least mypy) don't support. Mypy treats everything as deriving from `object`. If there is interest by other type checkers, we can later reconsider including this, but right now it feels quite specialized. 

For historical reasons there are some mypy-specific things in stdlib stubs, and I'd like to move away type checker specific things instead of adding more of them.

Reply:
My intention here isn't to provide pyright-specific functionality. Of course I can hard-code behaviors into pyright, but that defeats the purpose. My goal here was to provide a standardized way that all type checkers could leverage such that they could properly handle None without hard-coding behaviors.

If you don't like this proposal, can you offer other proposals that achieve this goal? Or do you think that the goal is misguided?

Reply:
Sorry if I was unclear. My concern is about what makes `_NoneType` special in that it can't be treated as a subclass of `object`. I don't understand why that would be needed and assumed that this would be specific to pyright, but I may be mistaken there. Can give more motivation why `_NoneType` is special in this way?

My alternative proposal would be to treat `_NoneType` as a subclass of `object`, since this seems to match how things work at runtime:

```
>>> type(None).mro()
[<class 'NoneType'>, <class 'object'>]
```

I'd also leave out all methods except `__bool__` and perhaps others such as `__repr__` that are not inherited from `object` (these seem to include `__subclasshook__`, `__new__` and `__init_subclass__`, but specifying these explicitly may not add any value).

Reply:
Thanks, that proposal works for me. There are some attributes that are defined for `object` that are not available on `None` (`__dict__`, `__slots__`, `__module__`), and the `__class__` property setter doesn't work for `None`, but those are edge cases that are probably fine to ignore.

Reply:
Note that `None` is hardly the only type that doesn't actually have `__dict__` or other attributes you mention. The actual semantics are just hard to model in the type system.